### PR TITLE
Update make_kn_lm.py

### PR DESCRIPTION
**Pull request by huangruizhe**
Fixed issue #163

### CODE REVIEW

**Comment by csukuangfj** [comment]
Is it ready for merge?

**Comment by huangruizhe** [comment]
> Is it ready for merge?

Not yet. Will let you know when it is done. I think @teowenshen  will test it on his dataset.

**Comment by teowenshen** [comment]
Hi @huangruizhe , thanks for the tag. 

The algorithm is stuck again at /i/. I think the conditions should be the other way round, i.e. assigning None when the denominator is zero.

Also, from my experience, some sum_z1_f_z summed up to more than 1.0, maybe due to Python's rounding, and yielded a large negative number. This caused print_as_arpa to throw a Math Domain Error at Line 353:

> `line += '\t{0}'.format('%.7f' % math.log10(bow))`

So maybe we should assign None to both when sum_z1_f_z is greater than or equal to 1.0?

File: icefall/shared/make_kn_lm.py [hunk 1]
```diff
@@ -241,7 +243,10 @@ def cal_bow(self):
                         for u in a_counts_for_hist.word_to_count.keys():  # Should be careful here: what is Z1
                             sum_z1_f_z += _counts_for_hist.word_to_f[u]
 
-                        counts_for_hist.word_to_bow[w] = (1.0 - sum_z1_f_a_z) / (1.0 - sum_z1_f_z)
+                        if 1.0 - sum_z1_f_z == 0:
```
Comment (danpovey) [inline icefall/shared/make_kn_lm.py:246]: Shouldn't this be < 0 instead of == 0?

Comment (huangruizhe) [reply icefall/shared/make_kn_lm.py]: Yes. Sorry for the mistake.

**Comment by huangruizhe** [comment]
@teowenshen Sorry... I made a mistake when making the PR. You are correct.

Regarding "some sum_z1_f_z summed up to more than 1.0", I agree with your guess.
I think the best way to guarantee the correctness (i.e. a valid probability distribution) is to check whether .

**Comment by teowenshen** [comment]
Hi @huangruizhe , I have just implemented your fix in my code. It ran smoothly and produced this end result, which seems to be as expected. I attach the end result here too for your inspection. 
 
[P.arpa.txt](https://github.com/k2-fsa/icefall/files/7801164/P.arpa.txt)

**Pull request merged**